### DISCUSSION TITLE
Thoughts on Using LPeg in Neovim and Regex Conversion

### DISCUSSION THREAD

**Discussion by brynne8** [neovim/neovim#29199] [General]
I’ve been diving into Neovim lately after noticing its scripts trending on GitHub. While I’m not a Neovim user myself and don’t have much experience with it, as a Lua enthusiast, I was excited to learn that Neovim bundles LPeg. This led me to explore some interesting possibilities.

While digging into the code, I noticed that Neovim uses LPeg for glob patterns. This struck me as a bit unusual since glob patterns are typically more straightforwardly converted to regex. It made me wonder about the decision-making process behind this choice. 

Additionally, I’ve seen that Neovim tends to favor Vim regex over PCRE, which got me thinking about the reasons for this preference. I came across some discussions hinting at the possibility of creating a converter between the two syntaxes. With this in mind, I decided to experiment with a simple converter from PCRE to Vim regex using LPeg. Here’s the toy converter I came up with:

```lua
local m = require('lpeg')
local re = require('re')

local cont = m.R("\128\191")
local any_utf8 = m.R("\0\127")
               + m.R("\194\223") * cont
               + m.R("\224\239") * cont * cont
               + m.R("\240\244") * cont * cont * cont

local pcre = re.compile([=[
  Expression <- {| Branch ('|' Branch)* |} -> merge_or
  Branch     <- {| Piece+ |} -> merge_seq
  Piece      <- {: Atom (Quantifier >> append)? :}
  Atom       <- {'.'} / Class / Complex / { EscOutside } / { [^|)] }
  Quantifier <- Range / '*?' -> '\{-}' / {'*'} / '+?' -> '\{-1,}' / '+' -> '\+' /
                '??' -> '\{-,1}' / '?' -> '\='
  Range      <- { '{' [0-9]+ (',' [0-9]*)? '}' }
  Complex    <- ( '(' ({ Special? } Expression) ')' ) -> handle_cplx
  Special    <- '?' (':' / '=' / '!' / '<=' / '<!' / '>')
  Class      <- {| {'[' '^'?} ClassItems {']'} |} -> merge_seq
  ClassItems <- (EscInside / !']' {%utf8} / {'-'})+
  EscOutside <- '\' %a ('{' [^}]* '}')?
  EscInside  <- '\' [dsw] -> esc_map / { '\' 'x' %x%x }
]=], {
  utf8 = any_utf8,
  esc_map = {
    d = '[:digit:]', s = '[:space:]', w = '0-9A-Za-z_'
  },
  merge_seq = function(t) return table.concat(t) end,
  merge_or = function(t) return table.concat(t, [[\|]]) end,
  append = function(a, b) return a .. b end,
  handle_cplx = function(special, expr)
    if special ~= '' then
      if special == '?:' then return ([[\%%\(%s\)]]):format(expr)
      elseif special == '?=' then return ([[\%%\(%s\)\@=]]):format(expr)
      elseif special == '?!' then return ([[\%%\(%s\)\@!]]):format(expr)
      elseif special == '?<=' then return ([[\%%\(%s\)\@<=]]):format(expr)
      elseif special == '?<!' then return ([[\%%\(%s\)\@<!]]):format(expr)
      elseif special == '?>' then return ([[\%%\(%s\)\@>]]):format(expr) end
    else
      return ([[\(%s\)]]):format(expr)
    end
  end
})

print(pcre:match([[a{3,6}\s[^az].z|abc]]))  -- a{3,6}\s[^az].z\|abc
print(pcre:match([[(?<=abc),\d+,(?=xyz)]])) -- \%\(abc\)\@<=,\d\+,\%\(xyz\)\@=
print(pcre:match([[abc[de\df]ghi]]))        -- abc[de[:digit:]f]ghi
```

It’s a toy converter, and I’m sure it’s not perfect. But it got me thinking about the potential for a more comprehensive tool. I wonder if anyone else has tackled similar projects or has insights into the broader context of regex usage within Neovim.

**Comment by clason** [comment]
Lua does not have regular expressions, it's as simple as that. Vim does, but crossing the Lua--Vim bridge incurs some overhead -- but not enough to warrant adding yet another regular expression engine (or conversion). Nobody has yet provided a compelling use case why the current two alternatives are not enough.

**Reply by mnemnion** [reply]
Lua does have regular expressions. It has LPeg, and PEGs are a generalized superset of regular expressions. The differences are merely syntactic: the semantics of LPeg are a superset of those supported by regular expressions. Since LPeg is a parser, converting a regex string to an LPeg pattern is a tractable goal. 

There's even the [re module](https://www.inf.puc-rio.br/~roberto/lpeg/re.html) which is part of LPeg itself, although Neovim doesn't appear to include it, which is too bad.

**Reply by clason** [reply]
That's a separate module, though -- and in that sense Lua has full regular expressions, too, via `vim.regex` (which is why we don't include "yet another regular expression module").

**Reply by mnemnion** [reply]
It's a separate module in a sense, given that it's [270 lines of pure Lua](https://github.com/roberto-ieru/LPeg/blob/master/re.lua) which just uses LPeg.  But not a separate module the way PCRE and LPeg are, it builds on LPeg, and is in fact used in the proof of concept at the top of this thread. 

Similarly, a file which translates regex syntax into LPeg patterns would be easy to write in pure Lua. The difference between this hypothetical translator, and `vim.regex`, is that `vim.regex` has to cross the abstraction boundary over into vim-land and back again, incurring the overhead which you referred to. 

LPeg does not have to do that. The notion of translating regex patterns into LPeg patterns is a solid one.  It doesn't have to go through the `re` module, of course, that's just a convenient shorthand. But it would just be a handful of lines of Lua.

**Reply by bfredl** [reply]
`vim.lpeg` and `vim.re` are both included. if you want to use these you can just use them directly already, commiting yet another a translation layer with its own set of quirks to core sounds like a ~~problem~~solution looking for a non-existent problem.

**Reply by mnemnion** [reply]
Oh how nice! I tried to import it with just `require("re")` out of habit. 

It's fine that you can't see the advantages of being able to translate regular expressions to a Lua-native pattern matching library built into an editor which makes heavy use of regular expressions and always has. We're not all equally gifted.
